### DISCUSSION TITLE
Vary amount of water delivered based on time elapsed after stimulus presented.

### DISCUSSION THREAD

**Discussion by msidikpr** [bonsai-rx/bonsai#1302] [Q&A] [answered]
Hello, 

I am having trouble implementing a change in my behavior task. The task is pretty simple in which a stimulus is presented a prompts a nose poke into an IR beam. The first break in the beam triggers water to be delivered by a solenoid. The amount delivered is controlled by the delay between the solenoid being turned on then off. I want to vary the amount of water delivered based on the time elapsed between the stim presentation and the first poke into the port. For example if the poke takes 1 second, more water is delivered than if the poke take 5 seconds from when the stimuli is presented. Workflow attached and thanks!           
[task_example.zip](https://github.com/bonsai-rx/bonsai/files/11055249/task_example.zip)

**Comment by bruno-f-cruz** [comment]
Couldn't compile your file, a lot of parts missing. Here's a possible solution:

![image](https://user-images.githubusercontent.com/7049351/227339369-50cabb87-9d25-4dcb-92ac-d081afc9f944.png)


`Take` the first output of the two events (`KeyDown` in the example, or pokes in your case), `Timestamp` them,  and pair them up in a `Zip` (this assumes that you only care about the first occurrence of either event). For each pair, calculate the difference between the timestamps and use that difference to create your delay value. On each `Zip` event, instantiate a `SelectMany` that turns on and off your valve with the variable `Delay`.

Hope it helps,
Bruno


[1302.zip](https://github.com/bonsai-rx/bonsai/files/11055586/1302.zip)

**Reply by msidikpr** [reply]
Hello again, thanks for the advice. I was able to use this method to get the total time of the trial. I am trying to use a python transform to instantiate the value I want to use for the valve from the trial time. I am not able to get the python transform to work and am assuming it is because the trial time value is in a Time.span object. I'm getting a scope storage object has no attribute 'process' error. Not sure how to get around this. 
![timed_task](https://user-images.githubusercontent.com/65372761/231257726-6d05c8f8-4ca7-411d-ab47-5b04b2189487.PNG)

**Reply by bruno-f-cruz** [reply]
I can't see what you wrote inside the `PythonTransform`. Can you add the workflow?

**Reply by msidikpr** [reply]
Here it is. 
![Capture](https://user-images.githubusercontent.com/65372761/231258574-a9e87632-5645-4d37-9358-a0e1bd79de18.PNG)
[Task.zip](https://github.com/bonsai-rx/bonsai/files/11203842/Task.zip)

**Reply by bruno-f-cruz** [reply, marked as answer]
If you really must use python:

```python
import clr
from System import TimeSpan

@returns(TimeSpan)
def process(value):
    if value >= 2:
        time = 1 
    elif value <= 2:
        time = 2
    return TimeSpan.FromSeconds(time)
```

Otherwise, you could do the same thing using native bonsai operators. Here's two possible solutions:

![image](https://user-images.githubusercontent.com/7049351/231261333-4b880040-7fdf-4863-8e4b-1492b8688e94.png)

<details>
    <summary>Workflow (Copy+Paste me)</summary>
    
```xml

<?xml version="1.0" encoding="utf-8"?>
<WorkflowBuilder Version="2.7.3"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:rx="clr-namespace:Bonsai.Reactive;assembly=Bonsai.Core"
                 xmlns:scr="clr-namespace:Bonsai.Scripting.Expressions;assembly=Bonsai.Scripting.Expressions"
                 xmlns="https://bonsai-rx.org/2018/workflow">
  <Workflow>
    <Nodes>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Timer">
          <rx:DueTime>PT0S</rx:DueTime>
          <rx:Period>PT2S</rx:Period>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Timestamp" />
      </Expression>
      <Expression xsi:type="MemberSelector">
        <Selector>Timestamp</Selector>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Timer">
          <rx:DueTime>PT0S</rx:DueTime>
          <rx:Period>PT1S</rx:Period>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Timestamp" />
      </Expression>
      <Expression xsi:type="MemberSelector">
        <Selector>Timestamp</Selector>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Zip" />
      </Expression>
      <Expression xsi:type="Subtract" />
      <Expression xsi:type="MemberSelector">
        <Selector>Seconds</Selector>
      </Expression>
      <Expression xsi:type="scr:ExpressionTransform">
        <scr:Expression>it &gt; 2 ? TimeSpan.FromSeconds(1) : TimeSpan.FromSeconds(2)</scr:Expression>
      </Expression>
      <Expression xsi:type="GreaterThan">
        <Operand xsi:type="IntProperty">
          <Value>2</Value>
        </Operand>
      </Expression>
      <Expression xsi:type="rx:Condition">
        <Workflow>
          <Nodes>
            <Expression xsi:type="WorkflowInput">
              <Name>Source1</Name>
            </Expression>
            <Expression xsi:type="WorkflowOutput" />
          </Nodes>
          <Edges>
            <Edge From="0" To="1" Label="Source1" />
          </Edges>
        </Workflow>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="TimeSpanProperty">
          <Value>PT1S</Value>
        </Combinator>
      </Expression>
      <Expression xsi:type="BitwiseNot" />
      <Expression xsi:type="rx:Condition">
        <Workflow>
          <Nodes>
            <Expression xsi:type="WorkflowInput">
              <Name>Source1</Name>
            </Expression>
            <Expression xsi:type="WorkflowOutput" />
          </Nodes>
          <Edges>
            <Edge From="0" To="1" Label="Source1" />
          </Edges>
        </Workflow>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="TimeSpanProperty">
          <Value>PT2S</Value>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Merge" />
      </Expression>
    </Nodes>
    <Edges>
      <Edge From="0" To="1" Label="Source1" />
      <Edge From="1" To="2" Label="Source1" />
      <Edge From="2" To="6" Label="Source1" />
      <Edge From="3" To="4" Label="Source1" />
      <Edge From="4" To="5" Label="Source1" />
      <Edge From="5" To="6" Label="Source2" />
      <Edge From="6" To="7" Label="Source1" />
      <Edge From="7" To="8" Label="Source1" />
      <Edge From="8" To="9" Label="Source1" />
      <Edge From="8" To="10" Label="Source1" />
      <Edge From="10" To="11" Label="Source1" />
      <Edge From="10" To="13" Label="Source1" />
      <Edge From="11" To="12" Label="Source1" />
      <Edge From="12" To="16" Label="Source1" />
      <Edge From="13" To="14" Label="Source1" />
      <Edge From="14" To="15" Label="Source1" />
      <Edge From="15" To="16" Label="Source2" />
    </Edges>
  </Workflow>
</WorkflowBuilder>

```

</details>
